### PR TITLE
fix parse_timestamp()

### LINKED ISSUE #26370: test-time-util is broken in certain timezones
### systemd version the issue has been seen with

main (253-rc2)

### Used distribution

Arch Linux/Ubuntu 20.04

### Linux kernel version used

Irrelevant

### CPU architectures issue was seen on

x86_64

### Component

tests

### Expected behaviour you didn't see

The tests should pass

### Unexpected behaviour you saw

Tests checking checking that conversion from timestamp to string and back sometimes fail in certain timezones.

### Steps to reproduce the problem

`TZ=Africa/Casablanca meson test -C build test-time-util --repeat 100`

From what it looks like, time util lib determines incorrectly whether the date _has_ belonged to the DST period or not and calculates the wrong timestamp most often missed by 1 hour.

The full list of broken timezones for Arch Linux with tzdata 2022g and glibc 2.36 is:
```
Africa/Casablanca
Africa/El_Aaiun
America/Argentina/Rio_Gallegos
America/Punta_Arenas
Antarctica/Casey
Antarctica/Davis
Antarctica/Palmer
Asia/Aqtau
Asia/Aqtobe
Asia/Atyrau
Asia/Baku
Asia/Bishkek
Asia/Oral
Asia/Qostanay
Asia/Qyzylorda
Asia/Samarkand
Asia/Srednekolymsk
Asia/Tbilisi
Asia/Yerevan
Chile/EasterIsland
Pacific/Easter
Turkey
```

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Thread 1 (Thread 0x7f10d20a6840 (LWP 783520)):
#0  0x00007f10d2a5864c in ?? () from /usr/lib/libc.so.6
#1  0x00007f10d2a08938 in raise () from /usr/lib/libc.so.6
#2  0x00007f10d29f253d in abort () from /usr/lib/libc.so.6
#3  0x00007f10d2e4fc70 in log_assert_failed (text=0x55b3a5a88ed0 "x / USEC_PER_SEC == y / USEC_PER_SEC", file=0x55b3a5a870e2 "src/test/test-time-util.c", line=398, func=0x55b3a5a89c00 <__func__.11> "test_FORMAT_TIMESTAMP") at ../src/basic/log.c:888
#4  0x000055b3a5a84620 in test_FORMAT_TIMESTAMP () at ../src/test/test-time-util.c:398
#5  0x000055b3a5a7f968 in run_test_table () at ../src/shared/tests.h:106
#6  0x000055b3a5a867fa in main (argc=1, argv=0x7ffe3c76ee38) at ../src/test/test-time-util.c:656
(gdb) frame 4
#4  0x000055b3a5a84620 in test_FORMAT_TIMESTAMP () at ../src/test/test-time-util.c:398
398                     assert_se(x / USEC_PER_SEC == y / USEC_PER_SEC);
(gdb) info locals
buf = 0x55b3a5c322c0 "Wed 2018-08-15 12:30:03 +01"
x = 1534332603388565
y = 1534336203000000
i = 91
__func__ = "test_FORMAT_TIMESTAMP"
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #26370.

### CODE REVIEW

File: src/basic/time-util.c [hunk 1]
```diff
@@ -868,12 +836,38 @@ static int parse_timestamp_impl(const char *t, usec_t *ret, bool with_tz) {
         if (weekday >= 0 && tm.tm_wday != weekday)
                 return -EINVAL;
 
+        /* gmtoff may be negative, and the date maye be too old (e.g. 1969-12-31 23:00:00 -06). */
+        if (gmtoff < 0 && tm.tm_year == 69 && tm.tm_mon == 11 && tm.tm_mday == 31) {
```
Comment (bluca) [inline src/basic/time-util.c]: why is this date hardcoded? could it not be tm_year=68 or so? would be good to have a test for this corner case

Comment (yuwata) [reply src/basic/time-util.c]: > why is this date hardcoded? could it not be tm_year=68 or so?

The hardcoded date is one day before the UNIX epoch (1969-12-31).
I extended the comment in the code why we need to handle the date explicitly.

> would be good to have a test for this corner case

There have already existed implicitly, but added more cases explicitly. PTAL.

File: src/basic/time-util.c [hunk 2]
```diff
@@ -890,24 +880,93 @@ static int parse_timestamp_impl(const char *t, usec_t *ret, bool with_tz) {
...
+        if (!timezone_is_valid(space + 1, LOG_DEBUG))
+                return parse_timestamp_maybe_with_tz(t, space - t, /* valid_tz = */ false, ret);
+
+        /* Shortcut. If the current $TZ is equivalent to the specified timezone, it is not necessary to fork
+         * the process. */
+        tz = getenv("TZ");
+        if (tz && *tz == ':' && streq(tz + 1, space + 1))
+                return parse_timestamp_maybe_with_tz(t, space - t, /* valid_tz = */ true, ret);
+
+        /* Otherwise, to make not pollute the current environment variables, let's fork the process and set
```
Comment (bluca) [inline src/basic/time-util.c]: "Otherwise, to avoid polluting the current..."

**Review by bluca — APPROVED** [review]

File: src/basic/time-util.c (hunk 2 above)
Comment (yuwata) [reply src/basic/time-util.c]: Thanks. Fixed.

**Comment by bluca** [comment]
@mrc0mmand is `TEST-10-ISSUE-2467` one of the known flaky ones?

**Comment by mrc0mmand** [comment]
> @mrc0mmand is `TEST-10-ISSUE-2467` one of the known flaky ones?

Yeah, that's definitely unrelated.

**Comment by bluca** [comment]
@yuwata I think this might be causing test failures: https://github.com/systemd/systemd/actions/runs/4264878769/jobs/7423526496

of course here they all passed, to make things easier...

File: src/basic/time-util.c [hunk 3]
```diff
@@ -890,24 +880,93 @@ static int parse_timestamp_impl(const char *t, usec_t *ret, bool with_tz) {
...
+        assert(len > 0);
+
+        buf = strndup(t, len);
+        if (!buf)
+                return -ENOMEM;
+
+        return parse_timestamp_impl(buf, /* with_tz = */ true, utc, isdst, gmtoff, ret);
+}
+
+static int parse_timestamp_maybe_with_tz(const char *t, size_t len, bool valid_tz, usec_t *ret) {
```
Comment (poettering) [inline src/basic/time-util.c:896]: hmm "len" is a bit of a misnomer, it's after all *not* length of `t`, but just some offset into it where the tz likely is...

maybe rename tz_offset? or tz_index? or tz_pos?

File: src/basic/time-util.c [hunk 4]
```diff
@@ -890,24 +880,93 @@ static int parse_timestamp_impl(const char *t, usec_t *ret, bool with_tz) {
...
+        if (streq(space + 1, "UTC"))
+                return parse_timestamp_with_tz(t, space - t, /* utc = */ true, /* isdst = */ -1, /* gmtoff = */ 0, ret);
+
+        /* If the timezone is compatible with RFC-822/ISO 8601 (e.g. +06, or -03:00) then parse the string as
+         * UTC and shift the result. */
+        k = strptime(space + 1, "%z", &tm);
+        if (k && *k == '\0') {
+                /* glibc accepts gmtoff more than 24 hours, but we refuse it. */
+                if ((usec_t) labs(tm.tm_gmtoff) > USEC_PER_DAY / USEC_PER_SEC)
+                        return -EINVAL;
```
Comment (poettering) [inline src/basic/time-util.c:952]: shouldn't this check better be placed in parse_timestamp_with_tz()? i.e. encapsulate it there?

**Comment by poettering** [comment]
sorry for being late for the review

**Comment by poettering** [comment]
(lgtm in general, just some nitpicks, see above)

**Comment by poettering** [comment]
hmm, this doesn't work so. for me tests don't pass anymore:

```
…
/* test_parse_timestamp */
Assertion 'usec >= expected' failed at src/test/test-time-util.c:663, function test_parse_timestamp_one(). Aborting.
Aborted
$ ls -al /etc/localtime 
lrwxrwxrwx. 1 root root 35 16. Sep 17:53 /etc/localtime -> ../usr/share/zoneinfo/Europe/Berlin
```

**Comment by bluca** [comment]
Disabling failing tests via https://github.com/systemd/systemd/pull/26612 for now

File: src/basic/time-util.c [hunk 5]
```diff
@@ -850,65 +824,149 @@ static int parse_timestamp_impl(const char *t, usec_t *usec, bool with_tz) {
...
         if (weekday >= 0 && tm.tm_wday != weekday)
                 return -EINVAL;
 
-        x = mktime_or_timegm(&tm, utc);
-        if (x < 0)
-                return -EINVAL;
+        if (gmtoff < 0) {
+                plus = -gmtoff * USEC_PER_SEC;
+
+                /* If gmtoff is negative, the string maye be too old to be parsed as UTC.
```
Comment (topimiettinen) [inline src/basic/time-util.c:844]: maye be -> may be

**Pull request merged**